`extract_array` can give results of different size

hamogu opened the pull request:
```
>>> import numpy as np
>>> from astropy.nddata.utils import extract_array
>>> large_array = np.arange(30).reshape(5,6)
>>> extract_array(large_array, (3, 4), (0, 0))
array([[0, 1],
       [6, 7]])
>>> extract_array(large_array, (3, 4), (3, 3))
array([[13, 14, 15, 16],
       [19, 20, 21, 22],
       [25, 26, 27, 28]])
```
It would be useful for me if we had either:

-    an option to return where the position I put in falls on the new image ((0,0) in the first example and (1,2) in the second example) or
-    an option to always return a (3,4) array, but fill missing values in some way (e.g. nan or return a mask of invalid entries) or
-    both (via a keyword). What do people think?

This was originally https://github.com/astropy/imageutils/issues/21 but got not resolved before merging with astropy, so I open it here again to raise visibility.

My use specific use case is PSF photometry (https://github.com/astropy/photutils/pull/159) where I would like to extract sub-arrays or a large image and perform the PSF photometry on those subarrays. Unfortunately, with the current behaviour, I cannot easily relate the pixel coordinates of the large array and the subarray. The `extract_array` routine has all the code to check if I am extracting something in the middle of an array or on the edges (that's how it decides how large the return array is), but there is no way to propagate this information back to the caller.

I think this is a problem in almost any application of `extract_array`, my PSF photometry is just one concrete example.

CC: @astrofrog @larrybradley

hamogu commented:
OK, I pushed code. I suspect that this will not make it into 1.0 and thus I have not written anything into the CHANGES.rst yet.

cdeil commented:
travis-ci failed because of this Sphinx issue:
https://travis-ci.org/astropy/astropy/jobs/48295532#L4253

I'm not convinced this change is the right approach. Maybe it would be better to have functions that compute nd slices and then the user can decide what to do (e.g. change in place, or copy with or without padding) ... maybe using other helper functions.

I don't have a concrete suggestion at the moment, but my suggestion would be to think about this and discuss some more, unless most people think it's important to make this change before the 1.0 release.

@adonath You're using `extract_array` in https://github.com/gammapy/gammapy/pull/234 ... what do you think?

hamogu commented:
@cdeil The function to compute ndslices is `overlap_arrays`. If a user wants just a slice, he/she can call that. All that `extract_array` does is to return `large_array[slices]` (padded with masked entries in my branch to ensure that the return array has the size requested).

hamogu commented:
I should point out that in the course of this PR, I found that the current implementation no only returns arrays or different sizes (which I consider a bug, but others may consider that a feature) it also gives clearly wrong results when called with invalid (negative) inputs (examples with current master, are fixed with the modified implementation in this PR):
```
>>> from astropy.nddata.utils import overlap_slices, extract_array
>>> l = np.arange(5)
>>> extract_array(l, (2,), (-5,))
array([0, 1])     # correct answer would be empty or (per discussion above) an exception
>>> extract_array(l, (2,), (-3,))
array([0, 1, 2, 3])   # incorrect and much longer than requested
>>> # It works again is the requested position is close to valid
>>> extract_array(l, (2,), (-2,))
array([], dtype=int32)
```

The problem is independent of the number of dimensions, so I chose 1d arrays in these example, but it is a problem for 2 d arrays, too:
```
>>> extract_array(la, (2,3), (-5,-4))
array([[0, 1, 2, 3]])
```

cdeil commented:
Ah, OK ... I didn't know [overlap_slices](http://astropy.readthedocs.org/en/latest/api/astropy.nddata.utils.overlap_slices.html) was part of the public API.

Still ... aren't there use cases where one does want to get smaller cutouts at the edged (and not have those padded or masked)? I think that's what we need in https://github.com/gammapy/gammapy/pull/234 .

Would it be bad to add an option to let the user control the behaviour at the large array edges where the small array doesn't completely fit?

hamogu commented:
When you use the current version of `extract_array` you get a cutout of unknown size. Let's say you have some X-ray (or gammy ray) event with a peak in pixel (x, y) and not you want to extract a 5*5 pixel island around it.
Let's say you determine the centroid of that and find it's at (0.5, 2.2) of the small array. How do you know what coordinate that is in the big array? Your code might try (x+0.5, y+2.2), but that is wrong at the edges.
Instead the code in gammypy would have to do:

    #event on the left side of array:
    if x < 5/2:
        x_event = x - (5 - small_array.shape[0]) / 2 + 0.5
    else:
        x_event = x + 0.5
    # event on left side of array:
    if y < 5/2:
        y_event = y - (5 - small_array.shape[1]) / 2 + 2.2
    else:
        y_event = y + 2.2

(Note that the example above does not include the correct rounding even/odd array sizes and that in practice 5, 0.5 and 2.2 would be variable names.) Does that not look a little more confusing than it should be?

I'm not opposed to offering a keyword to switch that behaviour (that's what I suggested at the top of this thread) but I've been convinced that it would be easier to support one and only one way to do it.

adonath commented:
@cdeil For https://github.com/gammapy/gammapy/pull/234 it doesn't make a difference because the boundary handling is done outside the `extract_array` function. In general, I guess, that's what most users will do anyway, because it depends a lot on the application what you want to do with the extracted 'incomplete' arrays at the boundaries. 

@hamogu  Sorry for not commenting earlier. I'm not in favor of returning a `MaskedArray`, because it could change the type of the array. When you extract a smaller array out of a larger one you would expect it to have exactly the same type as the larger one.

I would keep the default behaviour as it is.  That leaves is to the user, what to do with those 'incomplete' extracted arrays, because it think there are surely cases were you just don't care and discard them, as it was done in the photutils prototype . But I would add an optional boundary parameter for convenience. It could even be the same options as in `astropy.convolution.convolve` i.e. 'none', 'fill', 'wrap' and 'extend' or possibly 'reflect'. But that might be overkill... and padding with NaN is sufficient. If the smaller array is completely outside the boundaries an error is raised.

mhvk commented on astropy/nddata/utils.py:
```diff
@@ -42,11 +50,24 @@ def overlap_slices(large_array_shape, small_array_shape, position):
-                 zip(position, small_array_shape)]
-    edges_max = [int(pos + 0.5 + small_shape / 2.) for (pos, small_shape) in
-                 zip(position, small_array_shape)]
+    edges_min = [int(np.floor(pos + 0.5 - small_shape / 2.))
```
why the addition of `np.floor`? For positive numbers it does the same as `int`, while I'd think that negative positions would need special treatment.

hamogu replied:
This is required to avoid errors with negative positions.
If we want to allow extraction of partially overlapping arrays with a central position to the left of the larger array (such as extract an array of size 5 with a center at -1, which still overlaps for two positions such as shown below)

    ----0123456789
      sssss

The extracted array s would be `---01`.

hamogu replied:
Subject to the same caveat as my other comments: I thought "pos" would be the "position of the center" not the "index of the center". If that is indeed meant to be the index, then negative indices can be disallowed and there is no need for `np.floor`.

mhvk commented on astropy/nddata/utils.py:
```diff
@@ -11,6 +11,11 @@
            'overlap_slices']
 
 
+class NoOverlapError(ValueError):
```
Would it be more logical to base this on `IndexError`? (since that is what is given with `np.array([0])[1]`)

hamogu replied:
sure.

hamogu replied:
I only now realize that "pos" was probably meant to represent and index and not a coordinate. I've always thought of it as a coordinate, and I may have been wrong. That's why I chose "Value" and not "Index" for the error.

astrofrog commented:
Regarding the discussion re: maskedarray one can also return an array with NaN values outside the boundary. That could actually be a keyword argument if needed.

mhvk commented:
@hamogu - on the negative indices: yes, this is indeed inconsistent. In principle, there is no reason not to be able to give a negative offset -- meaning, as usual, counting from the far end of the array -- but the addition of the size should not be allowed to become positive. Still, the effort would likely not be worth the pain, so it may be easiest to just disallow negative indices.

hamogu commented:
What does a negative position mean?

I interpreted that as the position in a geometical sense (The array as indexes 0,1,2,3, ... ,so "-1" would be on to the left of "0"), but I now realize that this can also be interpreted as the "index of the position" (negative numbers simply mean it's counted form the right of the array.)

Which one was the intention of the original authors of this code?

If they intended the meaning "array indices" then I take everything back that I said and close this PR. In that case, I just suggest the simple addition of "**Index of the** Position of the small array's center, with respect to the large array." to the docstring.

mhvk commented:
@hamogu - for what it is worth, I think your interpretation is much more logical than mine: let's treat `pos` as a position, not an index, i.e., something that might be returned by a conversion from RA, Dec.

@adonath's point about keeping the array type the same is well-taken (especially since `MaskedArray` deals so poorly with subclasses). At the same time, I do like the idea that one always gets an array with the right shape back. Filling with `nan` would seem the best compromise.

larrybradley commented:
I'd second @astrofrog's suggestion that the fill value be a keyword parameter (e.g. with default=`np.nan`).  The user could then later create a `MaskedArray` for the boundary cases if needed.

adonath commented:
@hamogu My intention was `position` measured in pixel coordinates, rather than `index`, I just didn't care about negative positions. But as you suggested it should be handled in some way, because I think e.g. `wcs_world2pix` also returns negative pixel coordinates, when the world position is outside the image.

adonath commented:
Just to underline my comment above: I don't think `extract_array` should always return the given shape, because the information that the array was extracted from the boundary region can be useful for how the array data is processed later. E.g. as I've written above, you might just want to discard this cases or apply a renormalisation. My suggestion would be the same as @astrofrog, but additonally having the boundary handling optional with a parameter `boundary='fill'` and `'none'`. When the option is set to`'none'` the behaviour could be either as it is now (returning 'incomplete' arrays) or maybe better raising some `SmallerArrayIsNotFullyContained` error. `boundary='fill'` could be the default...

larrybradley commented:
Just thinking out loud here, but perhaps a better approach would be to make this a class (e.g. `DataCutout`?) that returns both a cutout array (which *would* be truncated at the edges) and a bunch of positional attributes (e.g. center, bounding box, slice tuple, etc.) relative to both the large and small images/data?

hamogu commented:
I see this more as a low-level function that could be used to build a classes like @larrybradley suggested on top of it. If we are talking e.g. about an image, that class could also handle changes to the WCS etc.  
On the other hand, in some classes (e.g. photometry) I probably want to use this operation in an inner loop and I really only want one return value (and as other comments point out above, even the padding with nans might already be too much for some use cases), so making this into a class might be more complicated than necessary.

hamogu commented:
The Travis failure us just a few sphinx warnings which I will fix when I add the changelog in the end.
Before that, I want to make sure that we agree that this is the way it should be implemented.

larrybradley commented:
A low-level function is good (as you say, a class can be built on top of it).  Thinking about this more, I do like @adonath's suggestion about having both options (trimmed vs. non-trimmed) available via a `boundary` keyword.  I'm working on a problem now where I actually do want the trimmed cutout arrays.  Have both options seems like a good way to go.

hamogu commented:
I implemented a "trim" option that returns smaller arrays as before this PR.
Unfortunately, I get a lot of other failures locally that have nothing to do with this, and I'll need to debug that before I can run the tests on my laptop, so there may still be Travis failures, but I push this up just in case.

hamogu commented:
I just pushed again to fix some of the sphinx errors, but know that this will fail again, because I cannot figure out how I can refer to the Exceptions that I defined in `utils.py`. You will see three different way that I thought where sensible (`IncompleteTableError`, `~astropy.nddata.utils.IncompleTableError` and  `nddata.utils.ImcompleteTableError`) but none of them works. 

Can anyone tell me how Sphinx will understand this?
Alternatively, I can rewrite the docstring and remove the references.

hamogu commented:
Anyone an idea how I can fix the sphinx failures?
(I don't know why the appveyor times out, but I suspect it's unrelated.)

embray commented on astropy/nddata/utils.py:
```diff
@@ -27,9 +35,18 @@ def overlap_slices(large_array_shape, small_array_shape, position):
+        array is sufficient. In the "strict" mode, the small array has to be
+        fully contained in the large array, otherwise an
+        `IncompleteOverlapError` is raised. In both modes,
+        non-overlapping arrays will raise a `~nddata.utils.NoOverlapError`.
```
I think this still needs to be fully spelled out as `~astropy.nddata.util.NoOverlapError`.  You might want to check any other references like it as well.

hamogu commented:
Thanks to @embray I have finally figured out how to make the links in sphinx (it works locally now).

I also rebased on master, and added a change log (the rebasing was needed because there was no 1.1 section in the change log before).

So, I ask you for a final round of comments so that we can get this done.

hamogu commented:
This has been "Ready-for-final-review" for about two weeks now. Anybody willing to do a final review?
Tests are all passing, only coverage/coveralls stalled (and I don't have permission to restart it).

astrofrog commented:
@hamogu - sorry for the delay! This looks good to me. Just a couple of comments/questions:

* What happens if the user passes floating point positions?
* Would it be conceptually easier to say that the position is the bottom left? (since the behavior is more intuitive for even-shaped arrays). I don't mind either way but just thought I'd raise the possibility.

@larrybradley @mhvk @cdeil @adonath - anyone else can review this one last time?

adonath commented:
@astrofrog I think the idea of this function was to have a simple utility function to quickly extract portions of your data for separate processing. I liked the idea of passing directly pixel coordinates (in contrast to interpret the position as an array index) and not to think about shapes etc. In my opinion (for the applications I see for this functions) one could actually disallow the extraction of even shaped arrays...

I think floating point positions are truncated by `np.floor`, which may not be the best behaviour, because:
```
extract_array(a, [5], [5.999])
```
Gives:
```
array([3, 4, 5, 6, 7])
```
I think floating positions should be rounded. @hamogu What do you think, would be reasonable?

Apart from that I only came across two very, very minor issues:

The default behaviour is a bit strange for integer arrays, because ´fill_value=np.nan´ doesn't work for integer arrays:
```
a = np.arange(10)
extract_array(a, [5], [0])
```
Currently returns: 
```
array([-2147483648, -2147483648,           0,           1,           2])
```
For the 1d case one maybe should allow integers as positions, which currently raises a ´TypeError´:
```
a = np.arange(10)
extract_array(a, [5], 5)
```

Another idea, that just came to my mind is to even allow lists of positions, to extract several parts of the array at the same time:
```
a = np.arange(10)
extract_array(a, [3], [2, 5])
```
Would return:
```
array([[2, 3, 4], [4, 5, 6]])
```
At the moment I don't see an application for this, but in general I think it could be used to rewrite parts of the psf photometry code to avoid python loops, but I guess that is future work...

hamogu commented:
@astrofrog: I don't like the notion of "bottom left" because this works in any number of dimensions and "bottom left" is only defined in two dimensions. Under no circumstances do we want to special case the axes in the second dimension. In my opinion, the most natural way to do is to follow the numpy convention of array indices. Besides, this PR only modifies an existing function. Changing the corner of the (0,0) position would be a non-backwards compatible change that would require going through a deprecation phase.

hamogu commented:
Lists of positions: I agree that extracting lists of positions is going to be a common use case. In fact, the whole purpose of the PR is that I need it for my long-standing PR in photutils to improve PSF photometry there. And it's not complicated to put a for loop somewhere to extract many arrays.
However, I'm not sure this simple helper function is the right place to put this for loop. I don't think there is a performance gain (but we can benchmark that later) because this function does not create a lot of objects, etc. Fundamentally, it relies on numpy indexing and that does not support the functionality we would need, so the loop has to be done in python. Plus, this open a whole other can or worms, e.g. what would the output format be, if not all the return slices have the same size? A list of numpy arrays?

For all these reasons I think it is more reasonable to write a loop that calls `extract_array` multiple times in the calling routine.

hamogu commented:
np.floor() vs. np.round()
----------------------------------
I used `np.floor` for backwards compatibility. The old implementation uses `int(x)` which does the same thing as `np.floor` for positive values of `x`.
You can think about it this way: (0,0) is the upper left corner of the first pixel and (0,0.9999) the upper right corner or the first pixel. In that convention 
`extract_array(a, [5], [5.0])` and `extract_array(a, [5], [5.999])` should give the same answer.

Of course, an alternative convention would be to define (0,0) as the center of the first pixel in the array. That's an easy change in the code, but before I do that, I want to make sure there is broad agreement for either convention.
For what it's worth, photutils uses (0,0) as the center, so I am inclined to do the change and use the same convention here.

Opinions?

In any case, I see that the floating point behaviour is not defined in the docs, so I will add that after we settle on (0,0) is the center or (0.5, 0.5) is the center of the first pixel.

hamogu commented:
Note that the last failure in appveyor seems unrelated to this code ("An exception has been raised that is likely due to a transient failure. If you are connecting to a SQL Azure database consider using SqlAzureExecutionStrategy.").

larrybradley commented:
+1 for leaving the position as the "center".  For even-sized shapes, I would assume this follows the same convention as the `size` parameter in `scipy.ndimage` filters, for example.

Also +1 for making `(0, 0)` the center of the first pixel in the array.  Offhand, I can't think of any convention that uses integer indices at the pixel corners.

hamogu commented:
Well, if I am not mistaken, the version in current master uses (0,0) as
the pixel corner. We can declare this a bug, and then I can fix it.
(I think the version in master was never meant to be used with floats,
so this choice was probably not intentional.)

On 03/16/2015 10:37 AM, Larry Bradley wrote:
>
> +1 for leaving the position as the "center". For even-sized shapes, I
> would assume this follows the same convention as the |size| parameter
> in |scipy.ndimage| filters, for example.
>
> Also +1 for making |(0, 0)| the center of the first pixel in the
> array. Offhand, I can't think of any convention that uses integer
> indices at the pixel corners.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/astropy/astropy/pull/3333#issuecomment-81704154>.
>

astrofrog commented:
Yes, I think should treat that as a bug and fix it so the center is (0,0)

mwcraig commented:
@hamogu -- looks like the appveyor fail was a fluke. Could you please make the bug fix you proposed when you get a chance?

hamogu commented:
I'm working on it. The problem is that `np.round` always rounds towards 0 as in 
```
In [16]: int(np.round(0.5))
Out[16]: 0

In [17]: int(np.round(-0.5))
Out[17]: 0
```
That's why I used `np.floor` originally to avoid a discontinuity at 0. Need to think at least one more night about this.

hamogu commented:
Anybody knows a better solution for this: https://github.com/astropy/astropy/pull/3333/files#diff-9d3c0ca15cfa1e9110d749db5d607202R24 ?
It seems to be an issue of round-off error. I'm running this doctest on travis now to see if it occurs in all numpy versions we support. Of course, it only happens at .5, but unfortunately we need numbers ending on .5 here a lot (every array with an odd dimension has coordinates like center +/- 1.5 pixels).

astrofrog commented:
That looks like a *weird* bug!

mhvk commented:
@hamogu - it seems to me numy is just following the standard definition of rounding, that it should be to the nearest *even* integer if the fraction is exactly 0.5 (this avoids biasing in rounding large samples). I think this also underlies the "rounding towards 0" that you observed (`np.round(-1.5)` yields `-2.0`).

Instead of adding our own `_offset` and `_round`, my suggestion would be to put a warning in the docstring to preferably use odd sizes of the to-be-extracted array (equivalent routines that I wrote used the half-width relative to the center position for this reason).

hamogu commented:
@mhvk: That's what I thought at first. But it's weirder than that:
```
>>> import numpy as np
>>> np.round([1.5, 2.5, 3.5, 4.5, 5.5, 6.5])
array([ 2., 2., 4., 4., 6., 6.])
```

mhvk commented:
See http://en.wikipedia.org/wiki/Rounding#Round_half_to_even -- it is "This is the default rounding mode used in IEEE 754 computing functions and operators." (and also called "Dutch rounding", which may explain why it was taught to me...).

hamogu commented:
In my current implementation, this is only an issue for even sized arrays. All tests for odd-sized arrays pass with the standard numpy `np.round` instead of my own `_round`. Thus, we can easily go with @mhvk's suggestion to remove my `_round` and `_offset`, but I would not be surprised if there was a hard-to-debug bug report about this at some point, because this introduces "jumps" when extracting even sized arrays where we get the same array twice and then "skip one":
```
In [7]: for i in np.arange(0,5): print extract_array(np.arange(5.), (2,), (i,))
[ 0.  1.]
[ 0.  1.]
[ 2.  3.]
[ 2.  3.]
[  4.  nan]
```
(Note that this is not run with the current branch, but a few commits back with `int(np.round( ...))` instead of `_round( ...)` .)

mhvk commented:
Would it be crazy to put out a warning for even-sized arrays?  The problem with any resolution is that the user will not know where the center falls in the extracted array anyway (unlike for odd sizes). Indeed, for even sizes it would seem the only sensible input for the center coordinates are integer + 0.5.

hamogu commented:
I think a warning is a good idea in any case.
Yet, extracting an e.g. 4*4 box and stepping over the image does not seem like such a crazy idea that astronomers would never try that (unless we actually raise an exception for even arrays). And in that case, users will expect that `center=i` and `center=i+1` differ by one step; though they might not care too much which convention we adopt (round up or round down) - for that part, I expect that most users will just try it out on the edge of the image and say "Ah, that's where I need to start my index".

astrofrog commented:
Just thinking out loud, but maybe whatever we do we should make sure that the user knows what the position they requested is in the new frame of reference of the array. That is, say I request a 10x10 cutout at a position of 57.2, 44.5 - what are the coordinates of that position in the new cutout? This is going to be essential for any kind of photometry application.

larrybradley commented:
@hamogu I really don't like that "jumping" behavior for even-sized arrays.  The `scipy` filters handle even-sized footprints fine, e.g. a simple example:

```
from scipy.ndimage import generic_filter
data = np.arange(5)
print(generic_filter(data, np.sum, (2,), mode='constant'))
[0 1 3 5 7]
```

So to be consistent, your example should return:
```
[np.nan, 0]
[0, 1]
[1, 2]
[2, 3]
[3, 4]
```

I don't see why we can't support even-sized extractions as long as we document which pixel is the "center".  Note that `generic_filter` has an `origin` keyword, where this can even be changed.

hamogu commented:
@larrybradley I don't like it either. I think it's very confusing. That's why my implementation does **not** use `np.round` but instead implements its own `_round`, so that the current state of the PR does **not** have those jumps.
We could add an `origin` keyword, but if there is not strong need, then I'd like to avoid that. My feeling is that astronomers will be fine with any convention we chose as long as it's documented. I smell feature-creep (we already introduced a keyword to pick how to handle the edges and another one for fill values).

larrybradley commented:
@hamogu I agree about not adding an `option` keyword.  Only the `generic_filter` seems to have the `origin` keyword.  The other pre-defined `scipy` filters don't have that option, but appear to assume `origin=0`, so I would recommend that should be the convention we use.

hamogu commented:
Also, I just noticed: This comment is the 59th comment on this PR, so we now have exactly one comment for every 4 lines of code. Can we try to bring this discussion to a close at some point?

mhvk commented:
@hamogu - just to be sure, while I'd prefer not to redefine what rounding means, I am not objecting to the approach you have taken -- as long as it is documented.

mwcraig commented:
@hamagou @astrofrog -- I think that is ready to merge once the travis error is fixed. I'm about to add line notes pointing out where the errors are, and caught one place where masking is still mentioned in the docstring.

mwcraig commented on astropy/nddata/utils.py:
```diff
@@ -20,16 +57,26 @@ def overlap_slices(large_array_shape, small_array_shape, position):
         Position of the small array's center, with respect to the large array.
         Coordinates should be in the same order as the array shape.
+        For a coordinate with an even number of elements, the position is
+        rounded up, e.g. extracting two elements with a center of `1` will
```
Need double backticks around 1

mwcraig commented on astropy/nddata/utils.py:
```diff
@@ -20,16 +57,26 @@ def overlap_slices(large_array_shape, small_array_shape, position):
         Coordinates should be in the same order as the array shape.
+        For a coordinate with an even number of elements, the position is
+        rounded up, e.g. extracting two elements with a center of `1` will
+        give positions `[0, 1]`.
```
Double backticks around [0, 1]

mwcraig commented on astropy/nddata/utils.py:
```diff
@@ -73,11 +150,26 @@ def extract_array(array_large, shape, position):
     array_small : `~numpy.ndarray`
-        The extracted array
+        The extracted array. Values that do not overlap with ``array_large``
+        are masked.
```
Instead of "are masked", maybe "are handled as described in mode"? Or just revert back to "The extracted array" and leave it at that?

astrofrog commented:
I agree that this is pretty much ready to merge apart from @mwcraig's small comments - sorry for letting this drag on. One final comment I want to re-iterate and which you had in your original description of this PR is that as a user, I want to know where the coordinates I gave lie in the new image. Can you add a couple of sentences to the docstring (under e.g. a ``Notes`` section) saying how I should compute that deterministically?

hamogu commented:
@larrybradley: For odd arrays we use the middle element as the center already, which, I believe, is exactly what you suggest. (I looked at both the scipy documentation and the scipy source code and it is not clear to me what happens for even sized arrays).

hamogu commented:
@astrofrog : If you want to know exactly what the coordinates of the input position are in terms of your new array, use `mode='partial'`. The position you put in will be exactly the center of the extracted array, with two complications:
- if you put in an float position instead of an int, then it might not be at the center of the pixel, but it's in the same pixel (e.g. if you put in `3.2`, the center of the new array will have the same coordinates as `3.0` in the original array)
- for even sized arrays it's rounded up, i.e. the "center" of a (2,2) array is the pixel (1,1), thus pixel (1,1) will have the same coordinates as the position you put in.

**However**, I understand that this is not necessarily sufficient. A user will want an option that returns the coordinates of position in the new array for even and odd shaped arrays and for any mode. Unfortunately, this is somewhat non-trivial in the presence of edges (and precisely this is the reason why I originally implemented the mode that we now call "partial", where it's easy). I think I will have to redesign `overlap_slices` to calculate these numbers. Sounds simple, but is a little tedious to get it right for even shaped arrays and rounding, so it won't happen today and maybe not even this weekend.

I'm sorry that now it's me to hold this up. I know that by now there are at least two other PRs that wait for this. Please bear with me.

larrybradley commented:
@hamogu  Yes, that's right.  I was just trying to clarify that the generic filter `origin=0` means the "middle", not index=0 (`origin` was probably not the best name).

BTW, for even-sized arrays, `scipy` appears to use (shape / 2) as the "center" index (as you have done here I think).

hamogu commented:
Travis is still running, but this is a heads up that I added a keyword to return the position in the coordinate system of the new array. Please review.

mwcraig commented on CHANGES.rst:
```diff
@@ -52,6 +52,9 @@ New Features
 
 - ``astropy.utils``
 
+  - extract array now offers different options to deal with array boundaries
```
Please move to the nddata section

mwcraig commented on astropy/nddata/utils.py:
```diff
@@ -20,16 +53,26 @@ def overlap_slices(large_array_shape, small_array_shape, position):
+        Shape of the large array (for 1D arrays, this can be an int or float).
+    small_array_shape : tuple or number
+        Shape of the small array (for 1D arrays, this can be an int or float).
+    position : tuple of integers or int
```
This isn't actually restricted to  integers, right? How about just `tuple or number` here, along with the minor edit a few lines down?

hamogu replied:
The shape of an array is restricted to an int. (No array has float dimensions), but he position is not. I changed this.

mwcraig commented on astropy/nddata/utils.py:
```diff
@@ -20,16 +53,26 @@ def overlap_slices(large_array_shape, small_array_shape, position):
+    position : tuple of integers or int
         Position of the small array's center, with respect to the large array.
         Coordinates should be in the same order as the array shape.
+        For a coordinate with an even number of elements, the position is
```
add: `Integer positions are at the pixel centers.`

mwcraig commented:
@hamagou -- only a couple  minor comments, thanks for adding the option to return position! If you don't have time to make these changes please speak up. They are not critical to merging this.

@astrofrog -- position is now returned,  haven't tried it extensively myself, but the test suite is pretty thorough. Unless you have concerns beyond my line comments, this looks good to merge; those line comments are not critical, can be fixed later.

hamogu commented:
Any more comments on my longest ever running PR?

pllim commented:
But it was created only in this year!

mwcraig commented:
This might not get merged until the  python in astronomy workshop is done.

Perhaps we could open a PR for a new unit? `1 hans = 1 comment/sloc` or `1 hans = 1 day-since-pr-opened/sloc` and any PR that reaches `0.5 hans` is automatically merged? :grin:

wkerzendorf commented:
@hamogu: @mwcraig and myself just looked through the comments and agree that this is good to merge. However, it can't be automatically merged. Can you rebase and ping us? Thanks for all your effort!

hamogu commented:
I get a merge conflict in **every single of 19 commits**. That's why I don't like PR's that are open this long (and if I remember right, this is about the forth time that I rebase).
I will close this PR and start from scratch from a fresh master checkout and just copy and paste the final changes in a text editor. That also cleans up the commit history, because this entire PR will only be a single commit.
(I understand that technically I could do that on the same branch, but I want to keep this branch in my repro in case anything goes wrong. )

hamogu commented:
@mwcraig @wkerzendorf : Please see #3727 which is this PR, with copy & paste of the changes into a new branch based of the current master.

embray commented:
> I get a merge conflict in every single of 19 commits.

Something about that doesn't sound right.  In the future I could probably try rebasing for you.
